Dart/src/com/jetbrains/lang/dart/analyzer/DartServerRootsHandler.java

```diff
@@ -108,6 +138,55 @@ private void updateRoots() {
     }
   }
 
+  boolean isInIncludedRoots(@Nullable final VirtualFile vFile) {
+    if (vFile == null) return false;
+
+    final String currentIncludedRoot = myIncludedRoots.size() > 0 ? myIncludedRoots.get(0) : "";
+
+    if (currentIncludedRoot == null || currentIncludedRoot.isEmpty()) return false;
+
+    if (myScopedAnalysisMode == DartProblemsViewSettings.ScopedAnalysisMode.All) {
+      final ProjectFileIndex fileIndex = ProjectRootManager.getInstance(myProject).getFileIndex();
+      if (fileIndex.isInLibraryClasses(vFile)) return true;
+
+      final Module module = fileIndex.getModuleForFile(vFile);
+      return (module != null && DartSdkLibUtil.isDartSdkEnabled(module));
+    }
+    else if (myScopedAnalysisMode == DartProblemsViewSettings.ScopedAnalysisMode.DartPackage) {
+      return FileUtil.toSystemDependentName(getEnclosingDartPackageDirectory(vFile)).equals(currentIncludedRoot);
+    }
+    return false;
+  }
+
+  @NotNull
+  public static String getEnclosingDirectory(@NotNull final VirtualFile vFile) {
+    return vFile.getParent().getPath();
+  }
+
+  @NotNull
+  public String getEnclosingDartPackageDirectory(@NotNull final VirtualFile vFile) {
+    // temp var to make sure that value is initialized
+    final VirtualFile packageRoot;
+
+    // TODO(jwren) follow up to extract method with duplicate code (from DartProblemsPresentationHelper)
+    final VirtualFile pubspec = Registry.is("dart.projects.without.pubspec", false)
+                                ? DartBuildFileUtil.findPackageRootBuildFile(myProject, myCurrentFile)
+                                : PubspecYamlUtil.findPubspecYamlFile(myProject, myCurrentFile);
```

Comment:
> You'll only need to tune the algorithm of searching pubspec/BUILD file a bit, as of now, it doesn't work for non-project files.

Reply:
I see yellow code here. Did you mean `vFile`?

BTW shouldn't `findPubspecYamlFile()` logic depend on the analysis mode as well? Maybe both these methods could check for the current analysis mode without explicitly passed parameter?